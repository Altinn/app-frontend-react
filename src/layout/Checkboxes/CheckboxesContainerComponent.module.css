/* Override design system component to enable horizontal layout
  No longer supported with horizontal layout in design system.
  Possible breaking change if we remove this option.
  */
.horizontal {
  display: flex;
  flex-direction: row;
  gap: var(--fds-spacing-2);
}

.checkBoxLabelContainer {
  display: flex;
  gap: var(--fds-spacing-1);
  align-items: center;
  font-size: 1rem;
}
