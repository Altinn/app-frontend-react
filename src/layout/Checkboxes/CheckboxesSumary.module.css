.formValue {
  font-weight: 500;
}

.labelWrapper {
  display: flex;
  align-items: center;
  gap: var(--fds-spacing-2);
}

.editButton {
  margin-left: auto;
  min-width: unset;
}
