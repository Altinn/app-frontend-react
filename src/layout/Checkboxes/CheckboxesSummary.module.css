.formValue {
  font-weight: 500;
}

.labelWrapper {
  display: flex;
  align-items: flex-start;
  gap: var(--fds-spacing-2);
}

.editButton {
  margin-left: auto;
  min-width: unset;
}
