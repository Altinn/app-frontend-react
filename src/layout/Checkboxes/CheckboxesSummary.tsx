import React from 'react';

import { Lang } from 'src/features/language/Lang';
import { useUnifiedValidationsForNode } from 'src/features/validation/selectors/unifiedValidationsForNode';
import { validationsOfSeverity } from 'src/features/validation/utils';
import { MultipleValueSummary } from 'src/layout/Summary2/CommonSummaryComponents/MultipleValueSummary';
import type { CheckboxSummaryOverrideProps } from 'src/layout/Summary2/config.generated';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';

type CheckboxesSummaryProps = {
  componentNode: LayoutNode<'Checkboxes'>;
  summaryOverrides?: CheckboxSummaryOverrideProps;
  displayData: string;
};

export const CheckboxesSummary = ({ componentNode, summaryOverrides, displayData }: CheckboxesSummaryProps) => {
  const maxStringLength = 75;
  const showAsList =
    summaryOverrides?.displayType === 'list' ||
    (!summaryOverrides?.displayType && displayData?.length >= maxStringLength);
  const validations = useUnifiedValidationsForNode(componentNode);
  const errors = validationsOfSeverity(validations, 'error');
  const title = componentNode.item.textResourceBindings?.title;
  return (
    <MultipleValueSummary
      title={<Lang id={title} />}
      errors={errors}
      componentNode={componentNode}
      showAsList={showAsList}
    />
  );
};
