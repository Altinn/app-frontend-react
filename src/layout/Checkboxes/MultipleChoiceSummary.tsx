import React from 'react';

import { Grid, List, ListItem, ListItemText, makeStyles } from '@material-ui/core';

import type { LayoutNode } from 'src/utils/layout/hierarchy';
import type { HComponent } from 'src/utils/layout/hierarchy.types';

export interface IMultipleChoiceSummaryProps {
  targetNode: LayoutNode<HComponent<'Checkboxes'>>;
}

const useStyles = makeStyles({
  list: {
    padding: 0,
  },
  listItem: {
    padding: 0,
  },
  // Match style in \src\components\summary\SingleInputSummary.tsx
  data: {
    fontWeight: 500,
    fontSize: '1.125rem',
    '& p': {
      fontWeight: 500,
      fontSize: '1.125rem',
    },
  },
});

export function MultipleChoiceSummary({ targetNode }: IMultipleChoiceSummaryProps) {
  const classes = useStyles();

  // PRIORITY: Find form data for component, check that it's string[]
  const formData: string[] = [];

  return (
    <Grid
      item
      xs={12}
      data-testid={'multiple-choice-summary'}
    >
      <List classes={{ root: classes.list }}>
        {formData &&
          Object.keys(formData).map((key) => (
            <ListItem
              key={key}
              classes={{ root: classes.listItem }}
            >
              <ListItemText
                id={key}
                primaryTypographyProps={{ classes: { root: classes.data } }}
                primary={formData[key]}
              />
            </ListItem>
          ))}
      </List>
    </Grid>
  );
}
