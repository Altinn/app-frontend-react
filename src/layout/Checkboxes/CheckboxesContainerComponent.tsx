import React from 'react';

import { Checkbox, HelpText } from '@digdir/design-system-react';
import cn from 'classnames';

import { AltinnSpinner } from 'src/components/AltinnSpinner';
import { OptionalIndicator } from 'src/components/form/OptionalIndicator';
import { RequiredIndicator } from 'src/components/form/RequiredIndicator';
import { DeleteWarningPopover } from 'src/components/molecules/DeleteWarningPopover';
import { useGetOptions } from 'src/features/options/useGetOptions';
import { useDelayedSavedState } from 'src/hooks/useDelayedSavedState';
import { useLanguage } from 'src/hooks/useLanguage';
import classes from 'src/layout/Checkboxes/CheckboxesContainerComponent.module.css';
import { shouldUseRowLayout } from 'src/utils/layout';
import { getPlainTextFromNode } from 'src/utils/stringHelper';
import type { PropsFromGenericComponent } from 'src/layout';

export type ICheckboxContainerProps = PropsFromGenericComponent<'Checkboxes'>;

const defaultSelectedOptions: string[] = [];

export const CheckboxContainerComponent = ({
  node,
  formData,
  isValid,
  handleDataChange,
  overrideDisplay,
}: ICheckboxContainerProps) => {
  const { id, layout, readOnly, textResourceBindings, required, labelSettings, alertOnChange } = node.item;
  const { lang, langAsString } = useLanguage();
  const {
    value: _value,
    setValue,
    saveValue,
  } = useDelayedSavedState(handleDataChange, formData?.simpleBinding ?? '', 200);

  const value = _value ?? formData?.simpleBinding ?? '';
  const selected = value && value.length > 0 ? value.split(',') : defaultSelectedOptions;
  const { options: calculatedOptions, isFetching } = useGetOptions({
    ...node.item,
    node,
    formData: {
      type: 'multi',
      values: selected,
      setValues: (values) => {
        setValue(values.join(','));
      },
    },
  });
  const [checkedItems, setCheckedItems] = React.useState<string[]>(selected);

  const handleChange = (checkedItems: string[], confirmedChange: boolean = false) => {
    setCheckedItems(checkedItems);
    const isBeingChecked = checkedItems.length > selected.length;
    // if alertOnChange is true, we need to wait for the user to confirm the change
    // checkedItems longer than selected means that the user has checked a new item
    // this logic lets the user check new items without triggering the alert
    if (!alertOnChange || confirmedChange || isBeingChecked) {
      const checkedItemsString = checkedItems.join(',');
      if (checkedItemsString !== value) {
        setValue(checkedItems.join(','));
      }
    }
  };

  const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
    // Only set value instantly if moving focus outside of the checkbox group
    if (!event.currentTarget.contains(event.relatedTarget)) {
      saveValue();
    }
  };

  const labelTextGroup = (
    <span className={classes.checkBoxLabelContainer}>
      {lang(node.item.textResourceBindings?.title)}
      <RequiredIndicator required={required} />
      <OptionalIndicator
        labelSettings={labelSettings}
        required={required}
      />
      {textResourceBindings?.help && (
        <HelpText title={langAsString(textResourceBindings?.help)}>{lang(textResourceBindings?.help)}</HelpText>
      )}
    </span>
  );
  const horizontal = shouldUseRowLayout({
    layout,
    optionsCount: calculatedOptions.length,
  });
  const ariaLabel = overrideDisplay?.renderedInTable ? langAsString(textResourceBindings?.title) : undefined;

  return isFetching ? (
    <AltinnSpinner />
  ) : (
    <div
      id={id}
      key={`checkboxes_group_${id}`}
      onBlur={handleBlur}
    >
      <Checkbox.Group
        className={cn({ [classes.horizontal]: horizontal })}
        legend={labelTextGroup}
        description={lang(textResourceBindings?.description)}
        disabled={readOnly}
        onChange={(values) => handleChange(values)}
        hideLegend={overrideDisplay?.renderLegend === false}
        error={!isValid}
        aria-label={ariaLabel}
        value={selected}
      >
        {calculatedOptions.map((option) => (
          <EnhancedCheckbox
            id={`${id}-${option.label.replace(/\s/g, '-')}`}
            key={option.value}
            description={lang(option.description) as string}
            value={option.value}
            checked={selected.includes(option.value)}
            alertOnChange={alertOnChange}
            renderedInTable={overrideDisplay?.renderedInTable}
            singleCheckbox={calculatedOptions.length === 1}
            label={option.label}
            helpText={option.helpText}
            handleChange={handleChange}
            checkedItems={checkedItems}
          />
        ))}
      </Checkbox.Group>
    </div>
  );
};

interface ICheckboxProps {
  id: string;
  description?: string;
  value: string;
  label: string;
  helpText?: string;
  checked: boolean;
  alertOnChange?: boolean;
  renderedInTable?: boolean;
  singleCheckbox?: boolean;
  handleChange: (checkedItems: string[], confirmedChange: boolean) => void;
  checkedItems: string[];
}

const EnhancedCheckbox = ({
  id,
  description,
  value,
  label,
  helpText,
  checked,
  alertOnChange,
  renderedInTable,
  singleCheckbox,
  handleChange,
  checkedItems,
}: ICheckboxProps) => {
  const [popoverOpen, setPopoverOpen] = React.useState(false);
  const { lang, langAsString } = useLanguage();
  const hideLabel = renderedInTable === true && singleCheckbox;

  const handleUncheck = () => {
    setPopoverOpen(false);
    handleChange(checkedItems, true);
  };
  const checkBox = (
    <Checkbox
      id={id}
      name={value}
      description={description}
      value={value}
      checked={checked}
      size='small'
      onChange={alertOnChange && checked ? () => setPopoverOpen(true) : undefined}
    >
      {
        <span className={cn({ 'sr-only': hideLabel }, classes.checkBoxLabelContainer)}>
          {langAsString(label)}
          {helpText && <HelpText title={getPlainTextFromNode(helpText)}>{lang(helpText)}</HelpText>}
        </span>
      }
    </Checkbox>
  );

  if (popoverOpen) {
    return (
      <DeleteWarningPopover
        trigger={checkBox}
        placement='left'
        deleteButtonText={langAsString('group.row_popover_delete_button_confirm')}
        messageText={langAsString('group.row_popover_delete_message')}
        onCancelClick={() => {
          setPopoverOpen(false);
        }}
        onPopoverDeleteClick={handleUncheck}
        open={popoverOpen}
        setOpen={setPopoverOpen}
      />
    );
  }
  return checkBox;
};
