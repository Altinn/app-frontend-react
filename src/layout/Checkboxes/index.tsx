import React, { forwardRef } from 'react';
import type { JSX } from 'react';

import { useLanguage } from 'src/features/language/useLanguage';
import { getCommaSeparatedOptionsToText } from 'src/features/options/getCommaSeparatedOptionsToText';
import { useNodeOptionsSelector } from 'src/features/options/useNodeOptions';
import { CheckboxContainerComponent } from 'src/layout/Checkboxes/CheckboxesContainerComponent';
import { CheckboxesSummary } from 'src/layout/Checkboxes/CheckboxesSummary';
import { CheckboxesDef } from 'src/layout/Checkboxes/config.def.generated';
import { MultipleChoiceSummary } from 'src/layout/Checkboxes/MultipleChoiceSummary';
import type { LayoutValidationCtx } from 'src/features/devtools/layoutValidation/types';
import type { DisplayDataProps } from 'src/features/displayData';
import type { IUseLanguage } from 'src/features/language/useLanguage';
import type { NodeOptionsSelector } from 'src/features/options/OptionsStorePlugin';
import type { PropsFromGenericComponent } from 'src/layout';
import type { SummaryRendererProps } from 'src/layout/LayoutComponent';
import type { CheckboxSummaryOverrideProps } from 'src/layout/Summary2/config.generated';
import type { LayoutNode } from 'src/utils/layout/LayoutNode';
import type { NodeFormDataSelector } from 'src/utils/layout/useNodeItem';

export class Checkboxes extends CheckboxesDef {
  render = forwardRef<HTMLElement, PropsFromGenericComponent<'Checkboxes'>>(
    function LayoutComponentCheckboxesRender(props, _): JSX.Element | null {
      return <CheckboxContainerComponent {...props} />;
    },
  );

  private getSummaryData(
    node: LayoutNode<'Checkboxes'>,
    langTools: IUseLanguage,
    optionsSelector: NodeOptionsSelector,
    nodeDataSelector: NodeFormDataSelector,
  ): { [key: string]: string } {
    const value = nodeDataSelector(node).simpleBinding ?? '';
    const { options } = optionsSelector(node);
    return getCommaSeparatedOptionsToText(value, options, langTools);
  }

  getDisplayData(
    node: LayoutNode<'Checkboxes'>,
    { langTools, optionsSelector, nodeFormDataSelector }: DisplayDataProps,
  ): string {
    return Object.values(this.getSummaryData(node, langTools, optionsSelector, nodeFormDataSelector)).join(', ');
  }

  renderSummary({ targetNode, nodeFormDataSelector }: SummaryRendererProps<'Checkboxes'>): JSX.Element | null {
    const langTools = useLanguage();
    const options = useNodeOptionsSelector();
    const summaryData = this.getSummaryData(targetNode, langTools, options, nodeFormDataSelector);
    return <MultipleChoiceSummary formData={summaryData} />;
  }

  renderSummary2(
    componentNode: LayoutNode<'Checkboxes'>,
    summaryOverrides?: CheckboxSummaryOverrideProps,
  ): JSX.Element | null {
    const displayData = this.useDisplayData(componentNode);
    return (
      <CheckboxesSummary
        componentNode={componentNode}
        displayData={displayData}
        summaryOverrides={summaryOverrides}
      />
    );
  }

  validateDataModelBindings(ctx: LayoutValidationCtx<'Checkboxes'>): string[] {
    return this.validateDataModelBindingsSimple(ctx);
  }
}
