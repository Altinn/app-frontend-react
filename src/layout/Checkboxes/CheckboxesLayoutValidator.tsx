import { useEffect } from 'react';

import { useLanguage } from 'src/features/language/useLanguage';
import { NodesInternal } from 'src/utils/layout/NodesContext';
import type { NodeValidationProps } from 'src/layout/layout';

export function CheckboxesLayoutValidator(props: NodeValidationProps<'Checkboxes'>) {
  const { node, externalItem } = props;
  const { langAsString } = useLanguage();
  const group = externalItem.dataModelBindings.group;
  const deletionStrategy = externalItem.deletionStrategy;
  const checkedBinding = externalItem.dataModelBindings.checked;

  const addError = NodesInternal.useAddError();

  useEffect(() => {
    let error: string | null = null;

    if (!group) {
      if (!!deletionStrategy || !!checkedBinding) {
        error = langAsString('config_error.deletion_strategy_no_save_to_list');
      }
    } else if (group) {
      if (!deletionStrategy) {
        error = langAsString('config_error.save_to_list_no_deletion_strategy');
      }
      if (deletionStrategy === 'soft' && !checkedBinding) {
        error = langAsString('config_error.soft_delete_no_is_deleted');
      }
      if (deletionStrategy === 'hard' && !!checkedBinding) {
        error = langAsString('config_error.hard_delete_with_is_deleted');
      }
    }

    if (error) {
      addError(error, node);
      window.logErrorOnce(`Validation error for '${node.id}': ${error}`);
    }
  }, [addError, node, deletionStrategy, checkedBinding, langAsString, group]);

  return null;
}
