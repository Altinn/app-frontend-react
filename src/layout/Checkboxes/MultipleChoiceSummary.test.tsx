import React from 'react';

import { render as rtlRender } from '@testing-library/react';

import { MultipleChoiceSummary } from 'src/layout/Checkboxes/MultipleChoiceSummary';
import type { IMultipleChoiceSummaryProps } from 'src/layout/Checkboxes/MultipleChoiceSummary';

describe('MultipleChoiceSummary', () => {
  test('MultipleChoiceSummary', () => {
    const { asFragment } = render();
    expect(asFragment()).toMatchSnapshot();
  });
});

function render(props: Partial<IMultipleChoiceSummaryProps> = {}) {
  const defaultProps: IMultipleChoiceSummaryProps = {
    formData: ['This is a text', 'This is another text'],
  };

  return rtlRender(
    <MultipleChoiceSummary
      {...defaultProps}
      {...props}
    />,
  );
}
