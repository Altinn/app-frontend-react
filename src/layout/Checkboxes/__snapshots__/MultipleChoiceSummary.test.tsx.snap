// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultipleChoiceSummary MultipleChoiceSummary 1`] = `
<DocumentFragment>
  <div
    style="display: none;"
  />
  <div
    class="col-xs-12 item"
    data-testid="multiple-choice-summary"
    style="display: block;"
  >
    <ul
      class="MuiList-root list MuiList-padding"
    >
      <li
        class="MuiListItem-root listItem MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
          id="some-key"
        >
          <span
            class="MuiTypography-root data MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            This is a text
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root listItem MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
          id="some-other-key"
        >
          <span
            class="MuiTypography-root data MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            This is another text
          </span>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
